video/out/opengl/hwdec_cuda.c

```diff
@@ -0,0 +1,287 @@
+    p->image_params = *params;
+    params->imgfmt = IMGFMT_NV12;
+    params->hw_subfmt = 0;
+
+    ret = CHECK_CU(cuCtxPushCurrent(p->cuda_ctx));
+    if (ret < 0) {
+        return ret;
+    }
+
+    gl->GenTextures(2, p->gl_textures);
+    for (int n = 0; n < 2; n++) {
+        gl->BindTexture(GL_TEXTURE_2D, p->gl_textures[n]);
+        GLenum filter = GL_NEAREST;
+        gl->TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, filter);
+        gl->TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, filter);
+        gl->TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+        gl->TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+        gl->TexImage2D(GL_TEXTURE_2D, 0, n == 0 ? GL_R8 : GL_RG8,
+                       params->w / (n + 1), params->h / (n + 1),
+                       0, n == 0 ? GL_RED : GL_RG, GL_UNSIGNED_BYTE, NULL);
+    }
+    gl->BindTexture(GL_TEXTURE_2D, 0);
+
+    gl->GenBuffers(2, p->gl_pbos);
+    for (int n = 0; n < 2; n++) {
+        gl->BindBuffer(GL_PIXEL_UNPACK_BUFFER_ARB, p->gl_pbos[n]);
+        gl->BufferData(GL_PIXEL_UNPACK_BUFFER_ARB,
+                       params->w * params->h / (n + 1),
+                       NULL, GL_STREAM_DRAW_ARB);
+        gl->BindBuffer(GL_PIXEL_UNPACK_BUFFER_ARB, 0);
+        ret = CHECK_CU(cuGLRegisterBufferObject(p->gl_pbos[n]));
+        if (ret < 0) {
+            goto error;
+        }
+    }
+
+ error:
+    eret = CHECK_CU(cuCtxPopCurrent(&dummy));
+    if (eret < 0) {
+        return eret;
+    }
+
+    return ret;
+}
+
+static void destroy(struct gl_hwdec *hw)
+{
+    struct priv *p = hw->priv;
+    GL *gl = hw->gl;
+    CUcontext dummy;
+
+    // Don't bail if any CUDA calls fail. This is all best effort.
+    CHECK_CU(cuCtxPushCurrent(p->cuda_ctx));
+    for (int n = 0; n < 2; n++) {
+        if (p->gl_pbos[n] > 0) {
+            CHECK_CU(cuGLUnregisterBufferObject(p->gl_pbos[n]));
+        }
+    }
+    CHECK_CU(cuCtxPopCurrent(&dummy));
+
+    gl->DeleteBuffers(2, p->gl_pbos);
+    gl->DeleteTextures(2, p->gl_textures);
+    for (int n = 0; n < 2; n++) {
+        p->gl_textures[n] = 0;
+        p->gl_pbos[n] = 0;
+    }
+
+    hwdec_devices_remove(hw->devs, &p->hwctx);
+}
+
+static int map_frame(struct gl_hwdec *hw, struct mp_image *hw_image,
+                     struct gl_hwdec_frame *out_frame)
+{
+    struct priv *p = hw->priv;
+    GL *gl = hw->gl;
+    CUcontext dummy;
+    CUdeviceptr cuda_data;
+    size_t cuda_size;
+    int ret = 0, eret = 0;
+
+    ret = CHECK_CU(cuCtxPushCurrent(p->cuda_ctx));
+    if (ret < 0) {
+        return ret;
+    }
+
+    // Luma Plane
+    gl->BindBuffer(GL_PIXEL_UNPACK_BUFFER_ARB, p->gl_pbos[0]);
+    ret = CHECK_CU(cuGLMapBufferObject(&cuda_data, &cuda_size, p->gl_pbos[0]));
+    if (ret < 0) {
+        goto error;
+    }
+
+    ret = CHECK_CU(cuMemcpy(cuda_data, (CUdeviceptr)hw_image->planes[0],
+                            hw_image->w * hw_image->h));
```

Comment:
Are you sure this is the correct stride?

Also in theory you could use a dummy mp_image and use its meta information to handle Y and UV planes in a uniform way, e.g. like the vaapi code: https://github.com/mpv-player/mpv/blob/master/video/out/opengl/hwdec_vaegl.c#L307